### PR TITLE
HHH-16281 Inconsistent Behaivor of L2 cache between Hibernate 5 and 6

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-16281

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/java/JavaTypedExpressible.java [hunk 1]
```diff
@@ -11,4 +11,9 @@
  */
 public interface JavaTypedExpressible<T> {
 	JavaType<T> getExpressibleJavaType();
+
+	default int extractHashCode(T disassembledValue) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/java/JavaTypedExpressible.java:15]: I think this is unnecessary. We should just call the `hashCode` method on the disassembled state directly.

**Comment by beikov** [comment]
During our last meeting we decided to use a different approach, which is to introduce a `addToCacheKey(MutableCacheKey)` on `JavaTypedExpressible` instead, which will add the constituent part and hash code to a `MutableCacheKey` directly. That way, we don't need to expose two methods and walk the model twice.

**Comment by beikov** [comment]
When you work on this, could you please also make sure that we use the same approach for https://github.com/hibernate/hibernate-orm/pull/6211 @dreab8 ?

**Comment by dreab8** [comment]
superseded by https://github.com/hibernate/hibernate-orm/pull/6279

**Pull request closed without merging**